### PR TITLE
PDP module

### PR DESCRIPTION
**Pull request by zahinSJI**
Benadryl automation report with PDP module: [Link](https://drive.google.com/file/d/1gfjuDlxlHr3iNOIUFJBXhLY40Yg3momE/view?usp=sharing)

### CODE REVIEW

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+    public ProductPage() {
+        PageFactory.initElements(driver, this);
+        actions = new Actions(driver);
+    }
+
+    /**
+     * Locators
+     */
+    @FindBy(css = "#tab-menu-8651")
+    WebElement prodHeaderMenu;
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: @zahinSJI make all the locators as private

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+import org.openqa.selenium.interactions.Actions;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.PageFactory;
+import org.testng.Assert;
+
+import com.jnj.actions.Action;
+import com.jnj.base.BaseClass;
+
+public class ProductPage extends BaseClass{
+    Actions actions;
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: @zahinSJI this is not needed. pls remove along with the initialization part.

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+     */
+    public void verifyProdOverview() {
+    	this.visitPDP();
+        extentInfoLog("Overview is displayed : ", Action.isDisplayed(driver, prodOverview));
+    }
+
+    /**
+     * Visits a product page from the homepage.
+     * Then, verifies the buttons.
+     */
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: @zahinSJI parameters are missing in the functions description

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+	WebElement closePrivacyBtn;
+    
+    public WebElement getJumpToHeadings(String headingText) {
+        return driver.findElement(By.xpath("//div[@class='jump-to-wrapper']//ul//li//a[text()='" + headingText + "']"));
+    }
+	
+	/**
+	 * Function to click on close privacy pop-up button
+	 */
+	public void closePrivacyPopup() {
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: @zahinSJI why this function is here? it is already there in the signup page.

File: BenadrylAutomation/src/test/java/com/jnj/testcases/ProductPageTest.java [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+import org.testng.SkipException;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import com.jnj.base.BaseClass;
+import com.jnj.dataproviders.ProductPageProvider;
+
+public class ProductPageTest extends BaseClass {
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/test/java/com/jnj/testcases/ProductPageTest.java:16]: @zahinSJI the format of the file doesn't look good. Pls format the file properly.

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java [hunk 6]
```diff
@@ -0,0 +1,313 @@
...
+        extentInfoLog("Image is displayed : ", Action.isDisplayed(driver, prodImg));
+    }
+
+    /**
+     * Visits a product page from the homepage.
+     * Then, verifies the title of the product.
+     */
+    public void verifyProdTitle() {
+    	this.visitPDP();
+        extentInfoLog("Title is displayed : ", Action.isDisplayed(driver, prodTitle));
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: @zahinSJI what is this??? where is assertion?? just checking displayed???

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+			Action.performActionwithExtentInfoLog(closePrivacyBtn, "click", "Clicking on : Close button");
+		} else extentFailLog("Privacy pop-up displayed : ", false);
+	}
+
+    /**
+     * Visits a product page from the homepage.
+     */
+    public void visitPDP() {
+        Action.performActionwithExtentInfoLog(prodHeaderMenu, "click", "Clicking on : " + prodHeaderMenu.getText());    
+        Action.performActionwithExtentInfoLog(prodTileTitle, "click", "Clicking on : " + prodTileTitle.getText());
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: @zahinSJI why is this static?? the product name should come from the data file

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+        extentInfoLog("Product title verified");
+    }
+
+    /**
+     * Visits a product page from the homepage.
+     * Then, verifies if the rating of the product is visible.
+     */
+    public void verifyProdRating() {
+    	this.visitPDP();
+        extentInfoLog("Rating is displayed : ", Action.isDisplayed(driver, prodRating));
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: @zahinSJI where is assertion??

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+    }
+
+    /**
+     * Visits a product page from the homepage.
+     * Then, verifies if the overview of the product is visible.
+     */
+    public void verifyProdOverview() {
+    	this.visitPDP();
+        extentInfoLog("Overview is displayed : ", Action.isDisplayed(driver, prodOverview));
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: @zahinSJI assertion?? just checking isdisplayed is not acceptable. i want assertion for each testcase.

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+     * Then, verifies the Write a Review button.
+     */
+    public void verifyWriteAReviewBtn(String expectedText) {
+    	this.visitPDP();
+        extentInfoLog("Button is displayed : ", Action.isDisplayed(driver, writeAReviewBtn));
+		String actualText= writeAReviewBtn.getText();
+        Assert.assertEquals(actualText, expectedText);
+		Action.performActionwithExtentInfoLog(writeAReviewBtn, "click", "Clicking on : " + actualText);
+        extentInfoLog("Modal is displayed : ", Action.isDisplayed(driver, writeAReviewModal));
+        extentInfoLog("Write a Review button verified");
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: @zahinSJI what is this?? you verifying like this? add assertions.

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java [hunk 11]
```diff
@@ -0,0 +1,313 @@
...
+     * Then, verifies the Buy Now button.
+     */
+    public void verifyBuyNowBtn(String expectedText) {
+    	this.visitPDP();
+        extentInfoLog("Button is displayed : ", Action.isDisplayed(driver, buyNowBtn));
+		String actualText = buyNowBtn.getText();
+        Assert.assertEquals(actualText, expectedText);
+		Action.performActionwithExtentInfoLog(buyNowBtn, "click", "Clicking on : " + actualText);
+        extentInfoLog("Modal is displayed : ", Action.isDisplayed(driver, buyNowModal));
+        extentInfoLog("Buy Now button verified");
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: @zahinSJI add assertion.

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java [hunk 12]
```diff
@@ -0,0 +1,313 @@
...
+    }
+
+    /**
+     * Visits a product page from the homepage.
+     * Then, verifies the jump to heading.
+     */
+    public void verifyJumpToOverview(String jumpToHeading, String sectionHeading) {
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: @zahinSJI all the below functions which starts with jumpto is not needed. You can just write 1 function for all. Modify accordingly.

File: BenadrylAutomation/src/test/java/com/jnj/testcases/ProductPageTest.java [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+        test = test.createNode(testcase);
+        if (execution.equalsIgnoreCase(defaultFlag)) {
+			selectEnv(runOn);
+            pdp.verifyButtons(expectedText);
+        } else {
+            throw new SkipException("Test skipped : " + testcase);
+        }
+    }
+
+    @Test(priority = 6, dataProvider = "jumpToOverview", dataProviderClass = ProductPageProvider.class)
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/test/java/com/jnj/testcases/ProductPageTest.java]: @zahin no need of separate methods. you can write just one common function for jumpTo. Also, you have created separate sheets, that is also not needed. Modify the code accordingly.

**Review by vaibhavNagvekarSJI — CHANGES_REQUESTED** [review]
Reviewed. Do the changes as mentioned in the comments

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java (hunk 1 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: Locators are made private as suggested

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java (hunk 4 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: Function removed from POM file for PDP

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java (hunk 2 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: Removed from all POM files

File: BenadrylAutomation/src/test/java/com/jnj/testcases/ProductPageTest.java (hunk 5 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/test/java/com/jnj/testcases/ProductPageTest.java:16]: Formatting updated

File: BenadrylAutomation/src/test/java/com/jnj/testcases/ProductPageTest.java (hunk 13 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/test/java/com/jnj/testcases/ProductPageTest.java]: Added as suggested

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java (hunk 12 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: Added as suggested

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java (hunk 3 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: Parameters are added

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java (hunk 11 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: Assertion added

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java (hunk 9 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: Assertion added

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java (hunk 7 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: Sheet data added

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java (hunk 6 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: Assertion added

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java (hunk 8 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: Assertion added

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java (hunk 10 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: Skipped due to shadow dom

**Comment by zahinSJI** [comment]
Here is the recently generated report: [Link](https://drive.google.com/file/d/1MIdCaam-qQbs9y2RZiVXu53NYLM-N0fJ/view?usp=sharing)

File: BenadrylAutomation/src/test/java/com/jnj/testcases/ProductPageTest.java (hunk 5 above)
Comment (vaibhavNagvekarSJI) [reply BenadrylAutomation/src/test/java/com/jnj/testcases/ProductPageTest.java:16]: fixed

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+	@FindBy(css = "#bv-product-reviews")
+	private WebElement reviewsSection;
+	
+	/**
+	 * Returns the WebElement of a 'Jump To' heading based on its text.
+	 *
+	 * @param headingText The text of the 'Jump To' heading.
+	 * @return WebElement The WebElement of the specified 'Jump To' heading.
+	 */
+	public WebElement getProductName(String path) {
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: @zahinSJI this are locators, make them private

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+	private WebElement faqsSection;
+
+	@FindBy(css = "#mini-panel-product_faq span")
+	private WebElement faqsHeading;
+
+	@FindBy(css = "#bv-product-reviews")
+	private WebElement reviewsSection;
+	
+	/**
+	 * Returns the WebElement of a 'Jump To' heading based on its text.
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: @zahinSJI no need to add description for locators

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java [hunk 16]
```diff
@@ -0,0 +1,284 @@
...
+	/**
+	 * Navigates to a product detail page from the homepage by clicking on the product header menu 
+	 * and a specific product title.
+	 * 
+	 * @param link of the product url
+	 */
+	public void visitDetailsPage(String path) {
+	    Action.performActionwithExtentInfoLog(prodHeaderMenu, "click", "Clicking on : " + prodHeaderMenu.getText());
+	    Action.performActionwithExtentInfoLog(getProductName(path), "click", "Clicking on : " + prodTileTitle.getText());
+	    Action.verifyPageUrl("products/" + path);
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: @zahinSJI why you sending text "products/" seperately??? send this from data file only

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java [hunk 17]
```diff
@@ -0,0 +1,284 @@
...
+
+	/**
+	 * Verifies that the product title is visible on the product detail page.
+	 * @param expectedTitle The text expected to be found on the title of the product page.
+	 */
+	public void verifyProdTitle(String expectedTitle) {
+	    this.visitPDP();
+	    String actualText = prodTitle.getText();
+	    Action.printAndAssert(actualText, expectedTitle);
+	    extentInfoLog("Title is displayed : ", Action.isDisplayed(driver, prodTitle));
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: @zahinSJI no need of this. remove this line.

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java [hunk 18]
```diff
@@ -0,0 +1,284 @@
...
+	}
+
+	/**
+	 * Verifies that the product rating is visible on the product detail page.
+	 */
+	public void verifyProdRating(String expectedRating) {
+	    this.visitPDP();
+	    String actualRating = prodRatingNum.getText();
+	    Action.printAndAssert(actualRating, expectedRating);
+	    extentInfoLog("Rating is displayed : ", Action.isDisplayed(driver, prodRating));
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: @zahinSJI no need of this. remove this line.

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java [hunk 19]
```diff
@@ -0,0 +1,284 @@
...
+
+	/**
+	 * Verifies that the product overview section is visible on the product detail page.
+	 * @param expectedOverview The text expected to be found on the overview being verified.
+	 */
+	public void verifyProdOverview(String expectedOverview) {
+	    this.visitPDP();
+	    String actualOverview = prodOverviewTxt.getText();
+	    Action.printAndAssert(actualOverview, expectedOverview);
+	    extentInfoLog("Overview is displayed : ", Action.isDisplayed(driver, prodOverview));
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: @zahinSJI no need of this. remove this line.

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java [hunk 20]
```diff
@@ -0,0 +1,284 @@
...
+	    this.visitPDP();
+	    extentInfoLog("Button is displayed : ", Action.isDisplayed(driver, buyNowBtn));
+	    String actualBtnText = buyNowBtn.getText();
+	    Action.printAndAssert(actualBtnText, expectedBtnText);
+	    Action.explicitWaitForElementTobeclickable(buyNowBtn, 30);
+	    Action.performActionwithExtentInfoLog(buyNowBtn, "click", "Clicking on : " + actualBtnText);
+	    Action.explicitWaitForElementTobeclickable(buyNowModalHeading, 30);
+	    String actualModalText = buyNowModalHeading.getText();
+	    Action.printAndAssert(actualModalText, expectedModalText);
+	    extentInfoLog("Buy Now modal is displayed : ", Action.isDisplayed(driver, buyNowModal));
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: @zahinSJI no need of this. remove this line.

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java [hunk 21]
```diff
@@ -0,0 +1,284 @@
...
+	            extentInfoLog("Section is displayed: ", Action.isDisplayed(driver, actualHeadingElement));
+	            return;
+	        default:
+	            throw new IllegalArgumentException("Invalid section type provided: " + jumpToHeading);
+	    }
+
+	    String actualHeadingText = actualHeadingElement.getText();
+	    extentInfoLog("Section is displayed : ", Action.isDisplayed(driver, actualHeadingElement));
+	    Action.printAndAssert(actualHeadingText, sectionHeading);
+	    extentInfoLog(jumpToHeadingText + " jump to heading verified");
```
Comment (vaibhavNagvekarSJI) [inline BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: @zahinSJI no need of this. remove this line.

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java (hunk 14 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: Done

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java (hunk 15 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: Descriptions removed for locators

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java (hunk 17 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: Removed

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java (hunk 18 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: Removed

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java (hunk 19 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: Removed

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java (hunk 20 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: Removed

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java (hunk 21 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: Removed

File: BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java (hunk 16 above)
Comment (zahinSJI) [reply BenadrylAutomation/src/main/java/com/jnj/pageobjects/ProductPage.java]: Changes made

**Review by vaibhavNagvekarSJI — APPROVED** [review]
Reviewed PDP module

**Pull request merged**